test/runtests.jl

```diff
@@ -12,8 +12,8 @@ using Zygote
 using Random, LinearAlgebra, Test
 
 using Distributions: meanlogdet
-using DistributionsAD: TuringUniform, TuringMvNormal, TuringMvLogNormal,
-    TuringPoissonBinomial
+using DistributionsAD: TuringUniform, #TuringMvNormal, TuringMvLogNormal,
+                       TuringPoissonBinomial
 using ForwardDiff: Dual
```

Comment:
I can't comment on the lines above, but the import of the AD backends (and hence the whole test setup) has to be changed. Currently all backends (ForwardDiff, ReverseDiff, Tracker, and Zygote) are loaded in this file, meaning that all optional adjoints will be loaded. Hence it is not tested if the optional AD support is actually separated properly.

Instead we have to make sure that only the optional dependencies for the Requires block are loaded that we want to test. One possibility would be to separate the AD tests into separate modules in separate files that are loaded depending on the AD backend, either manually or by using https://github.com/YingboMa/SafeTestsets.jl. More concretely, one would define the general test setup, distributions, etc in a function that takes the tested AD backend (as a symbol?) as an argument and depending on that tests the corresponding backend in `test_ad` instead of relying on the environment variable `AD` there.

Reply:
Yes, I very much agree! :) I'll take a look at SafeTestsets

Reply:
I tried separating the backend tests in a similar way as it is done in `tests/ad/utils.jl`. Not sure if its the best way though...